REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [3]:
import pandas as pd

In [4]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [5]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [6]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [125]:
len(bus)

6315

#### Question 2b: How many unique business IDs are there?  

In [126]:
len(bus['business_id'].unique())

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [127]:
bus['name'].value_counts()

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
                              ..
7-MISSION RESTAURANT           1
WONDERFUL DESSERT & CAFE       1
KITCHEN ISTANBAL               1
FAT BELLI DELI                 1
ALBA RAY'S                     1
Name: name, Length: 5692, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [128]:
# a summary of the dataframe, including zip code
bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

In [129]:
# Data type is an object, but let's examine a particular entry
type(bus.iloc[(1,5)])
#it is a string.

str

#### Question 3b

What are the unique values of postal_code?

In [130]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [7]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [8]:
vb = bus[bus['postal_code'].isin(validZip)]
vb

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
6310        NaN         NaN           NaN  
6311        NaN         NaN  +14154945048  
6312        NaN         NaN  +14158606036  
6313        NaN         NaN  +14153331878  
6314        NaN         NaN  +14158881393  

[5999 rows x 9 columns]

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [21]:
null_lon = vb[pd.isnull(vb['longitude'])]
len(null_lon)


2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [22]:
# dataframe with businesses without missing longitude values
not_null_lon = vb[pd.notnull(vb['longitude'])]
len(not_null_lon)

3516

In [23]:
#Counting number of businesses in each zipcode 
null_lon = null_lon['postal_code'].value_counts()
not_null_lon = not_null_lon['postal_code'].value_counts()
print(null_lon,not_null_lon)

94103    252
94110    251
94107    247
94102    196
94109    146
94133    135
94122    123
94111    115
94114    108
94118    100
94105    100
94124     88
94108     82
94115     76
94117     67
94132     65
94104     59
94112     59
94121     55
94123     53
94116     35
94134     29
94127     28
94131     14
Name: postal_code, dtype: int64 94110    334
94103    306
94133    282
94102    262
94109    237
94107    204
94111    167
94115    162
94122    147
94118    143
94108    142
94112    131
94117    130
94105    126
94114    122
94123    116
94121    106
94124     76
94104     74
94132     69
94116     58
94134     45
94127     43
94131     34
Name: postal_code, dtype: int64


In [27]:
#Merging the datasets
df = pd.merge(null_lon,not_null_lon, left_index=True, right_index=True)
df

#Renaming the columns
df.rename(columns = {'postal_code_x' : 'null_lon', 'postal_code_y' : 'not_null_lon'}, inplace = True)
df

null_lon  not_null_lon
94103       252           306
94110       251           334
94107       247           204
94102       196           262
94109       146           237
94133       135           282
94122       123           147
94111       115           167
94114       108           122
94118       100           143
94105       100           126
94124        88            76
94108        82           142
94115        76           162
94117        67           130
94132        65            69
94104        59            74
94112        59           131
94121        55           106
94123        53           116
94116        35            58
94134        29            45
94127        28            43
94131        14            34

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [29]:
#printing the proportion missing longitude
df["share"] = df["null_lon"]/ (df["null_lon"] + df["not_null_lon"]) *100
df

#top 5
df.nlargest(5, ['share'])


null_lon  not_null_lon      share
94107       247           204  54.767184
94124        88            76  53.658537
94132        65            69  48.507463
94114       108           122  46.956522
94122       123           147  45.555556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [30]:
#rows in the dataset
len(ins)

15430

In [31]:
#unique business
len(ins["business_id"].unique())

#Conclusion: there are multiple inspections per business

5730

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [32]:
# What values does type take on?
ins["type"].unique()

array(['routine', 'complaint'], dtype=object)

In [33]:
#How many occurrences of each value is in the DataFrame?
ins['type'].value_counts()


routine      15429
complaint        1
Name: type, dtype: int64

In [34]:
ins2 = ins['type'].value_counts()[ins['type'].value_counts()>=10]
ins2

routine    15429
Name: type, dtype: int64

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [35]:
ins['year'] = ins['date'].astype(str).str[0:4]
ins.head()


business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [36]:
#range of years covered
ins["year"].unique()

array(['2016', '2014', '2015', '2013'], dtype=object)

In [37]:
#number of inspections each year
ins["year"].value_counts()

2016    5840
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [38]:
#Dropping values records for any years with less than 50 inspections
ins3 = ins['year'].value_counts()[ins['year'].value_counts()> 50]
ins3

2016    5840
2014    5629
2015    3923
Name: year, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [39]:
#subsetting- year = 2016
ins_2016 = ins[ins['year'] == "2016"]
ins_2016.describe()

business_id        score          date
count   5840.000000  5840.000000  5.840000e+03
mean   44598.291438    90.233733  2.016068e+07
std    33999.580577     8.133708  2.998559e+02
min       10.000000    52.000000  2.016010e+07
25%     5619.750000    86.000000  2.016042e+07
50%    61064.500000    92.000000  2.016070e+07
75%    77445.750000    96.000000  2.016092e+07
max    89900.000000   100.000000  2.016122e+07

In [43]:
#inspections in 2016 per business
ins_2016.groupby(['business_id']).count()['score']

business_id
10       1
19       1
24       2
31       1
45       2
        ..
89515    1
89701    1
89790    1
89803    1
89900    1
Name: score, Length: 4543, dtype: int64

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [44]:
#Merging the datasets
df_merge = pd.merge(bus,ins_2016, how="left", on=["business_id"])
df_merge

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4              31      NORMAN'S ICE CREAM AND FREEZES   
...           ...                                 ...   
7607        90234                ACCESS SFUSD THE ARC   
7608        90236                      PHO DONG HUONG   
7609        90237                    WIN'S RESTAURANT   
7610        90238                    USA POWER MARKET   
7611        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
4              2801 LEAVENWORTH ST   San Francisco    CA       94133   
...                             ...            ...   ...         ...   
7607                1520 HOWARD ST   San Francisco    CA       94103   
7608             667 MONTEREY BLVD   San Francisco    CA       94127   
7609               3040 TARAVAL ST   San Francisco    CA       94116   
7610                1524 OCEAN AVE   San Francisco    CA       94112   
7611               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  score        date     type  year  
0     37.791116 -122.403816  +14154217044   82.0  20160503.0  routine  2016  
1     37.786848 -122.421547  +14157763262   94.0  20160513.0  routine  2016  
2     37.792888 -122.403135  +14156779494   98.0  20161005.0  routine  2016  
3     37.792888 -122.403135  +14156779494   96.0  20160311.0  routine  2016  
4     37.807155 -122.419004           NaN   84.0  20160816.0  routine  2016  
...         ...         ...           ...    ...         ...      ...   ...  
7607        NaN         NaN           NaN    NaN         NaN      NaN   NaN  
7608        NaN         NaN  +14154945048    NaN         NaN      NaN   NaN  
7609        NaN         NaN  +14158606036    NaN         NaN      NaN   NaN  
7610        NaN         NaN  +14153331878    NaN         NaN      NaN   NaN  
7611        NaN         NaN  +14158881393    NaN         NaN      NaN   NaN  

[7612 rows x 13 columns]

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [45]:
df_merge_20 = df_merge[["name", "address", "score"]].nsmallest(20, ['score'])
df_merge_20.sort_values(["score"], ascending=True) 


name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
7325                    POKI TIME     2101 LOMBARD ST    55.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
3563                  JADE GARDEN      3608 BALBOA ST    59.0
4234                        ITHAI         720 POST ST    59.0
6938         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
5013                 WOLFES LUNCH        1220 16TH ST    60.0
4850             HONG KONG LOUNGE     5322 GEARY BLVD    60.0
6151          VALLARTA'S TACO BAR     2022 MISSION ST    60.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!